### PR TITLE
[Merged by Bors] - intellij refactor

### PR DESCRIPTION
**Pull request by aviramha**

### CODE REVIEW

File: intellij-ext/build.gradle.kts [hunk 1]
```diff
@@ -37,6 +43,15 @@ intellij {
     plugins.set(properties("platformPlugins").split(',').map(String::trim).filter(String::isNotEmpty))
 }
 
+allprojects {
+    ext.jetbrains = [
+        version : "2022.2",
+    pycharm : "PythonCore:222.3345.40",
+    rubymine: "org.jetbrains.plugins.ruby:222.3345.16",
+    goland  : "org.jetbrains.plugins.go:222.3345.90",
+    scala   : "org.intellij.scala:2022.2.3"
```
Comment (eyalb181) [inline intellij-ext/build.gradle.kts]: What's happening here? What's the Scala bit? Please document

File: intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordEnabler.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                        MirrordSettingsState.versionCheckEnabled = true
+                        notification.expire()
+                    }
+                })
+                .addAction(object : NotificationAction("More info") {
+                    override fun actionPerformed(e: AnActionEvent, notification: Notification) {
+                        BrowserUtil.browse("https://github.com/metalbear-co/mirrord/blob/main/TELEMETRY.md")
+                    }
+                })
+                .addAction(object : NotificationAction("Deny (will disable version check as well)") {
```
Comment (eyalb181) [inline intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordEnabler.kt:52]: I'd remove the part in parentheses, it seems kind of insidious, as in "If you deny telemetries we'll also take away other functionality". 
I know that's not the case, but I think that could be the default interpretation for some users.

File: intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordApi.kt [hunk 3]
```diff
@@ -117,16 +93,31 @@ object MirrordApi {
...
+            // See if it's the final message
+            if (message.name == "mirrord preparing to launch"
+                && message.type == MessageType.FinishedTask) {
+                val success = message.success ?: throw Error("Invalid message")
+                if (success) {
+                    val innerMessage = message.message ?: throw Error("Invalid inner message")
+                    val executionInfo = gson.fromJson(innerMessage, MirrordExecution::class.java)
+                    return executionInfo.environment
+                } else {
+                    MirrordNotifier.errorNotification("mirrord failed launching", project)
```
Comment (eyalb181) [inline intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordApi.kt:105]:
```suggestion
                    MirrordNotifier.errorNotification("mirrord failed to launch", project)
```

File: intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordApi.kt [hunk 4]
```diff
@@ -117,16 +93,31 @@ object MirrordApi {
...
+                    throw Error("failed launching")
```
Comment (eyalb181) [inline intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordApi.kt:106]:
```suggestion
                    throw Error("failed to launch")
```

File: intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordEnabler.kt (hunk 2 above)
Comment (aviramha) [reply intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordEnabler.kt]: wdyt about current phrase?

File: intellij-ext/modules/products/pycharm/src/main/kotlin/com/metalbear/mirrord/products/pycharm/PythonRunConfigurationExtension.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+
+    override fun patchCommandLine(
+        configuration: AbstractPythonRunConfiguration<*>,
+        runnerSettings: RunnerSettings?,
+        cmdLine: GeneralCommandLine,
+        runnerId: String
+    ) {
+        val wsl = when (val request = createEnvironmentRequest(configuration, configuration.project)) {
```
Comment (eyalb181) [inline intellij-ext/modules/products/pycharm/src/main/kotlin/com/metalbear/mirrord/products/pycharm/PythonRunConfigurationExtension.kt:33]: Is this actually required in addition to the extendTargetEnvironment bit?

File: intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordExecManager.kt [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            return null
+        }
+        if (SystemInfo.isWindows && wslDistribution == null) {
+            MirrordNotifier.errorNotification("Can't use mirrord on Windows without WSL", project)
+            return null
+        }
+
+        var target: String? = null;
+        if (!MirrordConfigAPI.isTargetSet(project)) {
+            ApplicationManager.getApplication().executeOnPooledThread {
```
Comment (eyalb181) [inline intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordExecManager.kt:47]: Consider documenting why this is required

File: intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordEnabler.kt (hunk 2 above)
Comment (aviramha) [reply intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordEnabler.kt]: ?

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -283,6 +285,9 @@ jobs:
           distribution: zulu
           java-version: 11
       - run: cp LICENSE CHANGELOG.md intellij-ext/
+      - run: |
+          mkdir -p intellij-ext/bin/linux/arm64
```
[the remark this answers is not in the document]
Comment (aviramha) [reply .github/workflows/release.yaml:289]: fixed

File: intellij-ext/build.gradle.kts [hunk 8]
```diff
@@ -49,7 +79,11 @@ qodana {
 }
 
 tasks {
+    buildSearchableOptions {
```
[the remark this answers is not in the document]
Comment (aviramha) [reply intellij-ext/build.gradle.kts:82]: dunno but it required to work :\

File: mirrord/layer/src/hooks.rs [hunk 9]
```diff
@@ -74,6 +74,7 @@ impl<'a> HookManager<'a> {
             .or_else(|_| self.hook_any_lib_export(symbol, detour))
     }
 
+    #[cfg(target_os = "linux")]
```
[the remark this answers is not in the document]
Comment (aviramha) [reply mirrord/layer/src/hooks.rs:77]: No longer used now on other platforms

File: intellij-ext/modules/products/pycharm/src/main/kotlin/com/metalbear/mirrord/products/pycharm/PythonRunConfigurationExtension.kt (hunk 5 above)
Comment (aviramha) [reply intellij-ext/modules/products/pycharm/src/main/kotlin/com/metalbear/mirrord/products/pycharm/PythonRunConfigurationExtension.kt:33]: Those are two different extensions - I suppose there are cases where each used

File: intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordEnabler.kt (hunk 2 above)
Comment (eyalb181) [reply intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordEnabler.kt]: Personally I'd just remove the part in parentheses, but current phrase is fine

File: intellij-ext/build.gradle.kts (hunk 8 above)
[the remark this answers is not in the document]
Comment (eyalb181) [reply intellij-ext/build.gradle.kts:82]: Should add documentation then

[the remark this answers is not in the document]
Comment (aviramha) [reply intellij-ext/build.gradle.kts:82]: what would be the docs? I don't know what exactly it does and why we need it, it's gradle fun

[the remark this answers is not in the document]
Comment (eyalb181) [reply intellij-ext/build.gradle.kts:82]: "Removing this makes X stop working"

**Review by eyalb181 — APPROVED** [review]

**Pull request closed without merging**